packages/grid/x-data-grid/src/components/cell/GridEditSingleSelectCell.tsx

```diff
@@ -73,14 +74,23 @@ function GridEditSingleSelectCell(props: GridEditSingleSelectCellProps) {
   const baseSelectProps = rootProps.componentsProps?.baseSelect || {};
   const isSelectNative = baseSelectProps.native ?? false;
 
-  let valueOptions: Array<ValueOptions>;
-  if (typeof colDef.valueOptions === 'function') {
-    valueOptions = colDef.valueOptions!({ id, row, field });
+  let resolvedColumn: GridSingleSelectColDef | null = null;
+  if (isSingleSelectColDef(colDef)) {
+    resolvedColumn = colDef;
+  }
+
+  let valueOptions: Array<ValueOptions> | undefined;
+  if (typeof resolvedColumn?.valueOptions === 'function') {
+    valueOptions = resolvedColumn?.valueOptions({ id, row, field });
   } else {
-    valueOptions = colDef.valueOptions!;
+    valueOptions = resolvedColumn?.valueOptions;
   }
 
   const handleChange: SelectProps['onChange'] = async (event) => {
+    if (!valueOptions) {
+      return;
+    }
+
```

Comment:
```suggestion
  if (!valueOptions) {
    return null;
  }

  const handleChange: SelectProps['onChange'] = async (event) => {
```

Maybe avoid creating this callbacks at all if there's no valueOptions?